crates/js-backend/src/function.rs

```diff
@@ -0,0 +1,280 @@
+                }
+            }
+        }
+    }
+}
+
+impl HostImportFunction {
+    pub(crate) fn stored<'store>(&self, ctx: &JsContext<'store>) -> &'store StoredFunction {
+        &ctx.inner.functions[self.store_handle]
+    }
+
+    pub(crate) fn stored_mut<'store>(
+        &self,
+        ctx: JsContextMut<'store>,
+    ) -> &'store mut StoredFunction {
+        &mut ctx.inner.functions[self.store_handle]
+    }
+}
+
+// Safety: this is safe because its intended to run in single thread
+unsafe impl Send for HostImportFunction {}
+unsafe impl Sync for HostImportFunction {}
+unsafe impl Send for WasmExportFunction {}
+unsafe impl Sync for WasmExportFunction {}
+
+impl HostFunction<JsWasmBackend> for HostImportFunction {
+    fn new<F>(store: &mut impl AsContextMut<JsWasmBackend>, sig: FuncSig, func: F) -> Self
+    where
+        F: for<'c> Fn(&'c [WValue]) -> Vec<WValue> + Sync + Send + 'static,
+    {
+        let with_caller = move |_, args: &'_ [WValue]| func(args);
+        Self::new_with_caller(store, sig, with_caller)
+    }
+
+    fn new_with_caller<F>(
+        store: &mut impl AsContextMut<JsWasmBackend>,
+        sig: FuncSig,
+        func: F,
+    ) -> Self
+    where
+        F: for<'c> Fn(JsImportCallContext, &[WValue]) -> Vec<WValue> + Sync + Send + 'static,
+    {
+        // Safety: JsStoreInner is stored inside a Box and the Store is required by wasm-backend traits contract
+        // to be valid for function execution. So it is safe to capture this ptr into closure and deference there.
+        let store_inner_ptr = store.as_context_mut().inner as *mut JsStoreInner;
+        let enclosed_sig = sig.clone();
+        let wrapped = move |args: &js_sys::Array| -> js_sys::Array {
+            log::debug!(
+                "function produced by JsFunction:::new_with_caller call, signature: {:?}",
+                enclosed_sig
+            );
+
+            let store_inner = unsafe { &mut *store_inner_ptr };
+            let caller_instance = store_inner.wasm_call_stack.last().cloned().expect(
+                "Import cannot be called outside of an export call, when wasm_call_stack is empty",
+            );
+
+            let caller = JsImportCallContext {
+                store_inner,
+                caller_instance,
+            };
+
+            let args = wval_array_from_js_array(args, enclosed_sig.params().iter());
+            let result = func(caller, &args);
+            js_array_from_wval_array(&result)
+        };
+
+        let func =
+            Closure::wrap(Box::new(wrapped) as Box<dyn FnMut(&Array) -> Array>).into_js_value();
+
+        // Make a function that converts function args into array and wrap our func with it.
+        // Otherwise our closure will get only first argument.
+        let dyn_func = js_sys::Function::new_with_args(
+            "wrapped_func",
+            "return wrapped_func(Array.prototype.slice.call(arguments, 1))",
+        );
```

Comment:
I think that it also should be decoupled to a separate function to make code self-readable